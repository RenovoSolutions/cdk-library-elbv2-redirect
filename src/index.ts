import {
  aws_elasticloadbalancingv2 as elbv2,
} from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * The properties for the base redirect construct
 */
export interface RedirectProps {
  /**
   * The application load balancer this redirect applies to
   */
  readonly alb: elbv2.ApplicationLoadBalancer;

  /**
   * The source port to redirect from
   */
  readonly sourcePort: number;

  /**
   * The source protocol to redirect from
   */
  readonly sourceProtocol: elbv2.ApplicationProtocol;

  /**
   * The target port to redirect to
   */
  readonly targetPort: number;

  /**
   * The target protocol to redirect to
   */
  readonly targetProtocol: elbv2.ApplicationProtocol;
}

/**
 * A base redirect construct that takes source and destination ports and protocols
 *
 * Common use cases can be built from this construct
 */
export class Redirect extends Construct {
  constructor(scope: Construct, id: string, props: RedirectProps) {
    super(scope, id);

    props.alb.addListener(`redirect${props.sourcePort}to${props.targetPort}`, {
      port: props.sourcePort,
      protocol: props.sourceProtocol,
      defaultAction: elbv2.ListenerAction.redirect({
        port: props.targetPort.toString(),
        protocol: props.targetProtocol,
      }),
    });
  }
}

/**
 * Properties for the RedirectHttpHttps construct
 */
export interface RedirectHttpHttpsProps {
  /**
   * The application load balancer this redirect applies to
   */
  readonly alb: elbv2.ApplicationLoadBalancer;
}

/**
 * A construct that redirects HTTP to HTTPS for the given application load balancer
 */
export class RedirectHttpHttps extends Construct {
  constructor(scope: Construct, id: string, props: RedirectHttpHttpsProps) {
    super(scope, id);

    new Redirect(this, 'httpToHttps', {
      alb: props.alb,
      sourcePort: 80,
      sourceProtocol: elbv2.ApplicationProtocol.HTTP,
      targetPort: 443,
      targetProtocol: elbv2.ApplicationProtocol.HTTPS,
    });
  }
}
