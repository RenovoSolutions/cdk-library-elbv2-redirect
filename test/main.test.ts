import {
  App,
  Stack,
  aws_ec2 as ec2,
  aws_elasticloadbalancingv2 as elbv2,
} from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { Redirect, RedirectHttpHttps } from '../src/index';

test('Snapshot', () => {
  const app = new App();
  const stack = new Stack(app, 'test', {
    env: {
      account: '1234567890', // example
      region: 'us-east-1',
    },
  });

  const vpc = new ec2.Vpc(stack, 'vpc');

  const alb = new elbv2.ApplicationLoadBalancer(stack, 'alb', {
    internetFacing: true,
    vpc,
  });

  new Redirect(stack, 'redirect', {
    alb,
    sourcePort: 8080,
    sourceProtocol: elbv2.ApplicationProtocol.HTTP,
    targetPort: 8443,
    targetProtocol: elbv2.ApplicationProtocol.HTTPS,
  });

  new RedirectHttpHttps(stack, 'redirectHttpHttps', {
    alb,
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});