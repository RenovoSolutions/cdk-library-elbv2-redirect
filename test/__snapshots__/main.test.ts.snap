// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "alb8A8B13C2": {
      "DependsOn": [
        "vpcPublicSubnet1DefaultRoute10708846",
        "vpcPublicSubnet1RouteTableAssociation5D3F4579",
        "vpcPublicSubnet2DefaultRouteA1EC0F60",
        "vpcPublicSubnet2RouteTableAssociation21F81B59",
        "vpcPublicSubnet3DefaultRoute3F356A11",
        "vpcPublicSubnet3RouteTableAssociationD102D1C4",
      ],
      "Properties": {
        "LoadBalancerAttributes": [
          {
            "Key": "deletion_protection.enabled",
            "Value": "false",
          },
        ],
        "Scheme": "internet-facing",
        "SecurityGroups": [
          {
            "Fn::GetAtt": [
              "albSecurityGroup49866104",
              "GroupId",
            ],
          },
        ],
        "Subnets": [
          {
            "Ref": "vpcPublicSubnet1Subnet2E65531E",
          },
          {
            "Ref": "vpcPublicSubnet2Subnet009B674F",
          },
          {
            "Ref": "vpcPublicSubnet3Subnet11B92D7C",
          },
        ],
        "Type": "application",
      },
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
    },
    "albSecurityGroup49866104": {
      "Properties": {
        "GroupDescription": "Automatically created Security Group for ELB testalbDA5348D3",
        "SecurityGroupEgress": [
          {
            "CidrIp": "255.255.255.255/32",
            "Description": "Disallow all traffic",
            "FromPort": 252,
            "IpProtocol": "icmp",
            "ToPort": 86,
          },
        ],
        "SecurityGroupIngress": [
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 8080",
            "FromPort": 8080,
            "IpProtocol": "tcp",
            "ToPort": 8080,
          },
          {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow from anyone on port 80",
            "FromPort": 80,
            "IpProtocol": "tcp",
            "ToPort": 80,
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "albredirect8080to8443889EAB0D": {
      "Properties": {
        "DefaultActions": [
          {
            "RedirectConfig": {
              "Port": "8443",
              "Protocol": "HTTPS",
              "StatusCode": "HTTP_302",
            },
            "Type": "redirect",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "alb8A8B13C2",
        },
        "Port": 8080,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "albredirect80to443366A8C61": {
      "Properties": {
        "DefaultActions": [
          {
            "RedirectConfig": {
              "Port": "443",
              "Protocol": "HTTPS",
              "StatusCode": "HTTP_302",
            },
            "Type": "redirect",
          },
        ],
        "LoadBalancerArn": {
          "Ref": "alb8A8B13C2",
        },
        "Port": 80,
        "Protocol": "HTTP",
      },
      "Type": "AWS::ElasticLoadBalancingV2::Listener",
    },
    "vpcA2121C38": {
      "Properties": {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "vpcIGWE57CBDCA": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "vpcPrivateSubnet1DefaultRoute1AA8E2E5": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet1NATGateway9C16659E",
        },
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPrivateSubnet1RouteTableAssociation67945127": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet1RouteTableB41A48CC",
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet1Subnet934893E8",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPrivateSubnet1RouteTableB41A48CC": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPrivateSubnet1Subnet934893E8": {
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.96.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPrivateSubnet2DefaultRouteB0E07F99": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet2NATGateway9B8AE11A",
        },
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPrivateSubnet2RouteTable7280F23E": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPrivateSubnet2RouteTableAssociation007E94D3": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet2RouteTable7280F23E",
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet2Subnet7031C2BA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPrivateSubnet2Subnet7031C2BA": {
      "Properties": {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.128.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPrivateSubnet3DefaultRoute30C45F47": {
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": {
          "Ref": "vpcPublicSubnet3NATGateway82F6CA9E",
        },
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet3RouteTable24DA79A0",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPrivateSubnet3RouteTable24DA79A0": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPrivateSubnet3RouteTableAssociationC58B3C2C": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPrivateSubnet3RouteTable24DA79A0",
        },
        "SubnetId": {
          "Ref": "vpcPrivateSubnet3Subnet985AC459",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPrivateSubnet3Subnet985AC459": {
      "Properties": {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.160.0/19",
        "MapPublicIpOnLaunch": false,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PrivateSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPublicSubnet1DefaultRoute10708846": {
      "DependsOn": [
        "vpcVPCGW7984C166",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPublicSubnet1EIPDA49DCBE": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vpcPublicSubnet1NATGateway9C16659E": {
      "DependsOn": [
        "vpcPublicSubnet1DefaultRoute10708846",
        "vpcPublicSubnet1RouteTableAssociation5D3F4579",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet1EIPDA49DCBE",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet1Subnet2E65531E",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vpcPublicSubnet1RouteTable48A2DF9B": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPublicSubnet1RouteTableAssociation5D3F4579": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet1RouteTable48A2DF9B",
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet1Subnet2E65531E",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPublicSubnet1Subnet2E65531E": {
      "Properties": {
        "AvailabilityZone": "dummy1a",
        "CidrBlock": "10.0.0.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet1",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPublicSubnet2DefaultRouteA1EC0F60": {
      "DependsOn": [
        "vpcVPCGW7984C166",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPublicSubnet2EIP9B3743B1": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vpcPublicSubnet2NATGateway9B8AE11A": {
      "DependsOn": [
        "vpcPublicSubnet2DefaultRouteA1EC0F60",
        "vpcPublicSubnet2RouteTableAssociation21F81B59",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet2EIP9B3743B1",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet2Subnet009B674F",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet2",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vpcPublicSubnet2RouteTableAssociation21F81B59": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet2RouteTableEB40D4CB",
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet2Subnet009B674F",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPublicSubnet2RouteTableEB40D4CB": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPublicSubnet2Subnet009B674F": {
      "Properties": {
        "AvailabilityZone": "dummy1b",
        "CidrBlock": "10.0.32.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet2",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcPublicSubnet3DefaultRoute3F356A11": {
      "DependsOn": [
        "vpcVPCGW7984C166",
      ],
      "Properties": {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": {
          "Ref": "vpcIGWE57CBDCA",
        },
        "RouteTableId": {
          "Ref": "vpcPublicSubnet3RouteTableA3C00665",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "vpcPublicSubnet3EIP2C3B9D91": {
      "Properties": {
        "Domain": "vpc",
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet3",
          },
        ],
      },
      "Type": "AWS::EC2::EIP",
    },
    "vpcPublicSubnet3NATGateway82F6CA9E": {
      "DependsOn": [
        "vpcPublicSubnet3DefaultRoute3F356A11",
        "vpcPublicSubnet3RouteTableAssociationD102D1C4",
      ],
      "Properties": {
        "AllocationId": {
          "Fn::GetAtt": [
            "vpcPublicSubnet3EIP2C3B9D91",
            "AllocationId",
          ],
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet3Subnet11B92D7C",
        },
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet3",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "vpcPublicSubnet3RouteTableA3C00665": {
      "Properties": {
        "Tags": [
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "vpcPublicSubnet3RouteTableAssociationD102D1C4": {
      "Properties": {
        "RouteTableId": {
          "Ref": "vpcPublicSubnet3RouteTableA3C00665",
        },
        "SubnetId": {
          "Ref": "vpcPublicSubnet3Subnet11B92D7C",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "vpcPublicSubnet3Subnet11B92D7C": {
      "Properties": {
        "AvailabilityZone": "dummy1c",
        "CidrBlock": "10.0.64.0/19",
        "MapPublicIpOnLaunch": true,
        "Tags": [
          {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
          {
            "Key": "Name",
            "Value": "test/vpc/PublicSubnet3",
          },
        ],
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "vpcVPCGW7984C166": {
      "Properties": {
        "InternetGatewayId": {
          "Ref": "vpcIGWE57CBDCA",
        },
        "VpcId": {
          "Ref": "vpcA2121C38",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
